<a href="https://colab.research.google.com/github/heeeee-github/Stock_Analysis_in_Python_230327/blob/main/2%EC%A3%BC%EC%B0%A8/pandas_exercise.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [11]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

np.random.seed(0)

# 과제 1
## Q1. 


<div class="alert alert-success">
[문제1]     
    
insurance.csv'파일 읽어서 df에 할당하고, 상위 5개 항목 출력하기
</div>

------------

[힌트]

pd.read_csv()

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/zzhining/ml_basic/main/dataset/insurance.csv', index_col= False)
#df = pd.read_csv('insurance.csv', index_col= False)

df.head(5)

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

## Q2. 


<div class="alert alert-success">
[문제2]     
    
df의 요약(평균, 4분위수, 전체 갯수 등)정보 확인하기
</div>

------------

In [7]:
df.describe()

age          bmi     children       charges    debt
count  1338.000000  1338.000000  1338.000000   1338.000000  1338.0
mean     39.207025    30.663397     1.094918  13270.422265     0.0
std      14.049960     6.098187     1.205493  12110.011237     0.0
min      18.000000    15.960000     0.000000   1121.873900     0.0
25%      27.000000    26.296250     0.000000   4740.287150     0.0
50%      39.000000    30.400000     1.000000   9382.033000     0.0
75%      51.000000    34.693750     2.000000  16639.912515     0.0
max      64.000000    53.130000     5.000000  63770.428010     0.0

## Q3. 


<div class="alert alert-success">
[문제3]     
    
신규 컬럼(double bmi)을 생성하고, bmi의 2배가 되는 값을 할당하기
</div>

------------

In [8]:
df['double_bmi'] = df['bmi'] *2

df.head(5)

age     sex     bmi  children smoker     region      charges  debt  \
0   19  female  27.900         0    yes  southwest  16884.92400     0   
1   18    male  33.770         1     no  southeast   1725.55230     0   
2   28    male  33.000         3     no  southeast   4449.46200     0   
3   33    male  22.705         0     no  northwest  21984.47061     0   
4   32    male  28.880         0     no  northwest   3866.85520     0   

   double_bmi  
0       55.80  
1       67.54  
2       66.00  
3       45.41  
4       57.76

## Q4. 


<div class="alert alert-success">
[문제4]     
    
신규 컬럼(debt)을 생성하고, 0으로 채우기
</div>

------------

In [9]:
df['debt'] = 0
df.head(5)

age     sex     bmi  children smoker     region      charges  debt  \
0   19  female  27.900         0    yes  southwest  16884.92400     0   
1   18    male  33.770         1     no  southeast   1725.55230     0   
2   28    male  33.000         3     no  southeast   4449.46200     0   
3   33    male  22.705         0     no  northwest  21984.47061     0   
4   32    male  28.880         0     no  northwest   3866.85520     0   

   double_bmi  
0       55.80  
1       67.54  
2       66.00  
3       45.41  
4       57.76

## Q5. 


<div class="alert alert-success">
[문제5]     
    
debt에 0부터 값이 증가하도록 채우기
</div>

------------

In [14]:
# 강사코드
df['debt'] = np.arange(len(df))
df.head()

age     sex     bmi  children smoker     region      charges  debt  \
0   19  female  27.900         0    yes  southwest  16884.92400     0   
1   18    male  33.770         1     no  southeast   1725.55230     1   
2   28    male  33.000         3     no  southeast   4449.46200     2   
3   33    male  22.705         0     no  northwest  21984.47061     3   
4   32    male  28.880         0     no  northwest   3866.85520     4   

   double_bmi  
0       55.80  
1       67.54  
2       66.00  
3       45.41  
4       57.76

In [ ]:
for i in range(len(df['debt'])):
    df['debt'][i] = i  # 0부터 값이 증가하도록 채우기

df.head()


<ipython-input-34-eaf2ee5303aa>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['debt'][i] = i  # 0부터 값이 증가하도록 채우기


age     sex     bmi  children smoker     region      charges  double_bmi  \
0   19  female  27.900         0    yes  southwest  16884.92400       55.80   
1   18    male  33.770         1     no  southeast   1725.55230       67.54   
2   28    male  33.000         3     no  southeast   4449.46200       66.00   
3   33    male  22.705         0     no  northwest  21984.47061       45.41   
4   32    male  28.880         0     no  northwest   3866.85520       57.76   

   debt  
0     0  
1     1  
2     2  
3     3  
4     4

## Q6. 


<div class="alert alert-success">
[문제6]     
    
1번 인덱스의 정보 출력
</div>

------------

[힌트]
iloc



age                  18  
sex                male  
bmi               33.77  
children              1  
smoker               no  
region        southeast  
charges         1725.55  
double bmi        67.54  
debt                  1  

In [13]:
df.iloc[1]

age                  18
sex                male
bmi               33.77
children              1
smoker               no
region        southeast
charges       1725.5523
debt                  0
double_bmi        67.54
Name: 1, dtype: object

## Q7. 


<div class="alert alert-success">
[문제7]     
    
'double bmi'열 삭제하기
</div>

In [ ]:
# axis = 1 : 축을 컬럼방향으로 설정하기 위한 매개변수
# inplace = TRUE : df 객체에 있는 값을 연산 결과로 대체체
df.drop(['double_bmi'], axis=1)

## Q8. 


<div class="alert alert-success">
[문제8]     
    
 'age'열, 'sex'열만 출력하기
</div>

In [ ]:
df[['age','sex']]

age     sex
0      19  female
1      18    male
2      28    male
3      33    male
4      32    male
...   ...     ...
1333   50    male
1334   18  female
1335   18  female
1336   21  female
1337   61  female

[1338 rows x 2 columns]

# 과제2
## Q9.
<div class="alert alert-success">
[문제9]     
    
나이를 1살씩 증가시키기
</div>

In [15]:
# 강사코드
df['age'] = df['age'] + 1


In [16]:
df.head()

age     sex     bmi  children smoker     region      charges  debt  \
0   20  female  27.900         0    yes  southwest  16884.92400     0   
1   19    male  33.770         1     no  southeast   1725.55230     1   
2   29    male  33.000         3     no  southeast   4449.46200     2   
3   34    male  22.705         0     no  northwest  21984.47061     3   
4   33    male  28.880         0     no  northwest   3866.85520     4   

   double_bmi  
0       55.80  
1       67.54  
2       66.00  
3       45.41  
4       57.76

In [ ]:
for i in range(len(df['age'])):
    df['age'][i] = df['age'][i] + 1 # 0부터 값이 증가하도록 채우기

df.head()


<ipython-input-35-a5f3e22244e8>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['age'][i] = df['age'][i] + 1 # 0부터 값이 증가하도록 채우기


age     sex     bmi  children smoker     region      charges  double_bmi  \
0   20  female  27.900         0    yes  southwest  16884.92400       55.80   
1   19    male  33.770         1     no  southeast   1725.55230       67.54   
2   29    male  33.000         3     no  southeast   4449.46200       66.00   
3   34    male  22.705         0     no  northwest  21984.47061       45.41   
4   33    male  28.880         0     no  northwest   3866.85520       57.76   

   debt  
0     0  
1     1  
2     2  
3     3  
4     4

## Q10.
<div class="alert alert-success">
[문제10]     
    
charges의 표시형식을 소수점 아래 2자리로 바꾸기
</div>

[힌트]

함수 생성, 적용(apply/map)

In [18]:
# 강사 코드 1.

## round() : 반올림
df['charges'].round(2)
### 변경된 값을 적용하고 싶은 경우 : df['charges'] = df['charges'].round(2)
df.head()


age     sex     bmi  children smoker     region      charges  debt  \
0   20  female  27.900         0    yes  southwest  16884.92400     0   
1   19    male  33.770         1     no  southeast   1725.55230     1   
2   29    male  33.000         3     no  southeast   4449.46200     2   
3   34    male  22.705         0     no  northwest  21984.47061     3   
4   33    male  28.880         0     no  northwest   3866.85520     4   

   double_bmi  
0       55.80  
1       67.54  
2       66.00  
3       45.41  
4       57.76

In [19]:
# 강사 코드 2. format문 이용
test_num = 1234.5678
format(test_num , ".2f")

'1234.57'

In [20]:
# 강사 코드 3. format문 이용2
test_num = 1234.5678
def sosu(data) : 
  return format(data,".2f")

sosu(1245672.456)

'1245672.46'

In [22]:
df['charges'].apply(sosu)

0       16884.92
1        1725.55
2        4449.46
3       21984.47
4        3866.86
          ...   
1333    10600.55
1334     2205.98
1335     1629.83
1336     2007.94
1337    29141.36
Name: charges, Length: 1338, dtype: object

In [23]:
df['charges'] = df['charges'].apply(sosu)

df.head()

age     sex     bmi  children smoker     region   charges  debt  double_bmi
0   20  female  27.900         0    yes  southwest  16884.92     0       55.80
1   19    male  33.770         1     no  southeast   1725.55     1       67.54
2   29    male  33.000         3     no  southeast   4449.46     2       66.00
3   34    male  22.705         0     no  northwest  21984.47     3       45.41
4   33    male  28.880         0     no  northwest   3866.86     4       57.76

## Q11.
<div class="alert alert-success">
[문제11]     
    
컬럼 정렬(알파벳 순서대로) : age, bmi, charges... 순으로 정렬
</div>

[Hint]
sort_index()


In [25]:
df.sort_index(axis=1, inplace = True)
df.head()

age     bmi   charges  children  debt  double_bmi     region     sex smoker
0   20  27.900  16884.92         0     0       55.80  southwest  female    yes
1   19  33.770   1725.55         1     1       67.54  southeast    male     no
2   29  33.000   4449.46         3     2       66.00  southeast    male     no
3   34  22.705  21984.47         0     3       45.41  northwest    male     no
4   33  28.880   3866.86         0     4       57.76  northwest    male     no

## Q12.
<div class="alert alert-success">
[문제12]     
    
값 정렬(age) : 나이 순으로 정렬
</div>

[Hint]
sort_values


In [26]:
df.sort_values(by = ['age'])

age     bmi   charges  children  debt  double_bmi     region     sex  \
1248   19  39.820   1633.96         0  1248       79.64  southeast  female   
482    19  31.350   1622.19         0   482       62.70  southeast  female   
492    19  25.080   2196.47         0   492       50.16  northeast  female   
525    19  33.880  11482.63         0   525       67.76  southeast  female   
529    19  25.460   1708.00         0   529       50.92  northeast    male   
...   ...     ...       ...       ...   ...         ...        ...     ...   
398    65  25.600  14988.43         2   398       51.20  southwest    male   
335    65  34.500  13822.80         0   335       69.00  southwest    male   
378    65  30.115  16455.71         3   378       60.23  northwest  female   
1265   65  23.760  26926.51         0  1265       47.52  southeast    male   
635    65  38.190  14410.93         0   635       76.38  northeast    male   

     smoker  
1248     no  
482      no  
492      no  
525      no  
529      no  
...     ...  
398      no  
335      no  
378      no  
1265    yes  
635      no  

[1338 rows x 9 columns]

## Q13.
<div class="alert alert-success">
[문제13]     
    
나이('age') 평균구하기
</div>

[Hint]
mean

In [27]:
df['age'].mean()

40.20702541106129

## Q14.
<div class="alert alert-success">
[문제14]     
    
보험료('charges')와 가장 상관관계가 높은 변수는?
</div>

[Hint]
 - 공분산: df.cov()
 - 상관계수: df.corr()

In [ ]:
df.cov()

age           bmi    children       charges  \
age           197.401387      9.362337    0.719303  5.087480e+04   
bmi             9.362337     37.187884    0.093795  1.464730e+04   
children        0.719303      0.093795    1.453213  9.926742e+02   
charges     50874.802298  14647.304426  992.674197  1.466524e+08   
double_bmi     18.724673     74.375767    0.187590  2.929461e+04   
debt         -170.829095    -85.224706   11.747569 -1.578122e+04   

              double_bmi           debt  
age            18.724673    -170.829095  
bmi            74.375767     -85.224706  
children        0.187590      11.747569  
charges     29294.608853  -15781.223645  
double_bmi    148.751534    -170.449413  
debt         -170.449413  149298.500000

In [ ]:
df.corr()

age       bmi  children   charges  double_bmi      debt
age         1.000000  0.109272  0.042469  0.299008    0.109272 -0.031467
bmi         0.109272  1.000000  0.012759  0.198341    1.000000 -0.036169
children    0.042469  0.012759  1.000000  0.067998    0.012759  0.025221
charges     0.299008  0.198341  0.067998  1.000000    0.198341 -0.003373
double_bmi  0.109272  1.000000  0.012759  0.198341    1.000000 -0.036169
debt       -0.031467 -0.036169  0.025221 -0.003373   -0.036169  1.000000

# 과제 3
## Q15.
<div class="alert alert-success">
[문제15]     
    
 남자, 여자가 각각 몇명있는지 구하기
</div>

[Hint]
pd.value_counts


## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가 3명인 데이터 출력하기
</div>

## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가  5명 이상인 데이터 출력하기
</div>

## Q17.
<div class="alert alert-success">
[문제17]     
    
 bmi가  17 미만인 데이터 출력하기
</div>

# 과제 4 
## Q18.
<div class="alert alert-success">
[문제18]     
    
성별에 따른 보험료 평균
</div>

## Q19.
<div class="alert alert-success">
[문제19]     
    
흡연여부에 따른 보험료 평균
</div>

## Q20.
<div class="alert alert-success">
[문제20]     
    
지역별 보험료 평균
</div>

## Q21.
<div class="alert alert-success">
[문제21]     
    
흡연여부 + 지역별 데이터 평균
</div>

## Q22.
<div class="alert alert-success">
[문제22]     
    
 흡연여부 + 지역별 데이터 평균에서 'charges' 만 츨력하기
</div>

## Q23.
<div class="alert alert-success">
[문제23]     
    
 흡연여부 + 지역별 데이터 크기(size)
</div>

## Q24.
<div class="alert alert-success">
[문제24]     
    
smoker 여부에 따라, group_val 컬럼에 '흡연자', '비흡연자' 표시하기
</div>